add API and concept docs for Vault's Agent Registry and OAuth Resource Server Config profiles (#2485)
hashicorp/web-unified-docs: content/vault/v2.x/content/api-docs/secret/agent-registry.mdx

---
layout: api
page_title: 'Agent Registry - HTTP API'
description: This is the API documentation for managing agent registrations in Vault.
---

# Agent Registry

@include 'alerts/private-preview.mdx'

@include 'alerts/enterprise-only.mdx'

> The `/agent-registry` endpoint is used to manage agent registrations in Vault.

Reviewer: I think we should explain what's an agent registration is here

Author: The rest of the paragraph already explains the essence of what an agent registration is/does. I'd be happy to elaborate the explanation, but it would help if you can share something specific you wanted to see there.

Also, we could insert a link to the concepts document (after I write it), as an alternative strategy to giving a deeper explanation. However, I strongly agree with providing a mini explanation of what an agent registration is inline -- I dislike when docs make the reader chase links all over just to make a coherent story that could have been written as a few lines in place.

Agent registrations associate an identity entity with a set of ceiling policies
that constrain the maximum permissions available to an agent. An entity must

> have an agent registration before it can authenticate using OAuth credentials.

Reviewer 2: ```suggestion
The `/agent-registry` endpoint registers agent identities with a set of ceiling policies
that constrain the maximum permissions available to the agent. An entity must
have an agent registration before it can authenticate using OAuth credentials.
```
We can simplify the first two sentences. Also, does "agent" in this case refer to AI agents or something else? If it's AI, we should use "AI agents" the first time to clarify before simplifying to "agent"

Author: Thanks for calling this out. For technical clarity: in this feature, agent means an identity enabled for on-behalf-of workflows. That can include AI agents, but the mechanism is not limited to AI-specific behavior. I will update first mention to AI agent per docs guidance, then use agent afterward for readability.

## Register an agent

> This endpoint creates or updates an agent registration.

Reviewer 2: ```suggestion
Create or update an agent registration.
```

| Method | Path                      |
| :----- | :------------------------ |
| `POST` | `/agent-registry/register` |

### Parameters

- `id` `(string: "")` – ID of an existing registration. If set, updates the
  corresponding registration. If unset, creates a new agent registration.

- `display_name` `(string: <required>)` – Human-readable name of the
  registration. The name must be unique within the
  [target namespace](/vault/docs/enterprise/namespaces#vault-api-and-namespaces)
  for the request.

> - `entity_id` `(string: <required>)` – Entity ID representing this agent. The

Reviewer 2: ```suggestion
- `entity_id` `(string: <required>)` – Entity ID representing the agent. The
```

  entity must already exist in the identity store. Each entity can have only one
  agent registration per namespace.

- `description` `(string: "")` – Human-readable description of the
  registration.

- `owner` `(string: "")` – Owner of the registration.

- `ceiling_policies` `(list of strings: [])` – List of policy names applied

>   as ceiling policies to this agent registration. These define the maximum

Reviewer 2: ```suggestion
  as ceiling policies to the agent registration. The policies define the maximum
```

  permissions available to the agent. Cannot contain `root`.

- `no_default_ceiling_policy` `(bool: false)` – When set to `true`, opts out

>   of automatically adding the `default` and `default-ceiling` policies to this

Reviewer 2: ```suggestion
  of automatically adding the `default` and `default-ceiling` policies to the
```

  agent registration.

### Sample payload

```json
{
  "display_name": "my-agent",
  "entity_id": "fe2a8568-91cc-7580-40a4-6bed9a5867fa",
  "description": "Production deployment agent",
  "owner": "platform-team",
  "ceiling_policies": ["agent-deploy", "agent-secrets-read"]
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \
    http://127.0.0.1:8200/v1/agent-registry/register
```

### Sample response

```json
{
  "data": {
    "id": "a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6",
    "display_name": "my-agent"
  }
}
```

## Read registration by ID

> This endpoint queries an agent registration by its identifier.

Reviewer 2: ```suggestion
Fetch agent registration details by the registration ID.
```

| Method | Path                              |
| :----- | :-------------------------------- |
| `GET`  | `/agent-registry/registration/id/:id` |

### Parameters

- `id` `(string: <required>)` – Identifier of the registration.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    http://127.0.0.1:8200/v1/agent-registry/registration/id/a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6
```

### Sample response

```json
{
  "data": {
    "id": "a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6",
    "display_name": "my-agent",
    "entity_id": "fe2a8568-91cc-7580-40a4-6bed9a5867fa",
    "description": "Production deployment agent",
    "owner": "platform-team",
    "ceiling_policies": ["agent-deploy", "agent-secrets-read", "default", "default-ceiling"],
    "no_default_ceiling_policy": false,
    "creation_time": "2026-05-15T10:30:00Z",
    "last_updated_time": "2026-05-15T10:30:00Z"
  }
}
```

## Update registration by ID

> This endpoint updates an existing agent registration by its identifier.

Reviewer 2: ```suggestion
Update an existing agent registration using the registration ID.
```

- `entity_id` `(string: <required>)` – Entity ID representing this agent.

- `description` `(string: "")` – Human-readable description of the
  registration.

- `owner` `(string: "")` – Owner of the registration.

- `ceiling_policies` `(list of strings: [])` – List of policy names applied
  as ceiling policies to this agent registration. Cannot contain `root`.

- `no_default_ceiling_policy` `(bool: false)` – When set to `true`, opts out
  of automatically adding default ceiling policies.

### Sample payload

```json
{
  "display_name": "my-agent-renamed",
  "entity_id": "fe2a8568-91cc-7580-40a4-6bed9a5867fa",
  "description": "Updated production deployment agent",
  "owner": "platform-team",
  "ceiling_policies": ["agent-deploy", "agent-secrets-read", "agent-kv-read"]
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \
    http://127.0.0.1:8200/v1/agent-registry/registration/id/a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6
```

### Sample response

```json
{
  "data": {
    "id": "a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6",
    "display_name": "my-agent-renamed"
  }
}
```

## Delete registration by ID

> This endpoint deletes an agent registration by its identifier.

Reviewer 2: ```suggestion
Delete an agent registration by its registration ID.
```

| Method   | Path                              |
| :------- | :-------------------------------- |
| `DELETE` | `/agent-registry/registration/id/:id` |

### Parameters

- `id` `(string: <required>)` – Identifier of the registration to delete.

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request DELETE \
    http://127.0.0.1:8200/v1/agent-registry/registration/id/a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6
```

## List registrations by ID

> This endpoint returns a list of agent registrations by their identifiers.

Reviewer 2: ```suggestion
Fetch a list agent registrations by the associated agent ID.
```
TBH, the current endpoint descriptions are a little confusing. I'm guessing from context that this endpoint uses the ID of the agent (i.e., an agent may have multiple registrations associated with it) while the others use the ID of the registration since you're targeting a specific registration. But that seems wildly inconsistent 🤔

Author: Good catch. There is some nonsense in here somewhere (in my writing, not your feedback!). I'll take a deep look tomorrow.

Author: You're right—the wording was ambiguous, thank you for catching it. I reworked the endpoint descriptions so each one states exactly which field it matches on (`id`, `display_name`, or `entity_id`), and added an intro paragraph plus a "one registration per entity" note that should clear up the relationship between agents, registrations, and entities. Also aligned a few things with the style guide along the way. Curious whether the new phrasing reads cleanly to you—you'll catch things I won't.

| Method | Path                                        |
| :----- | :------------------------------------------ |
| `LIST` | `/agent-registry/registration/id`           |
| `GET`  | `/agent-registry/registration/id?list=true` |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request LIST \
    http://127.0.0.1:8200/v1/agent-registry/registration/id
```

### Sample response

```json
{
  "data": {
    "keys": [
      "a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6",
      "b7c19832-44ad-6fa1-9e23-1a5d0c8e47b3"
    ],
    "key_info": {
      "a3b04076-e7d9-251f-4a7d-8c4f0a7b59c6": {
        "display_name": "my-agent"
      },
      "b7c19832-44ad-6fa1-9e23-1a5d0c8e47b3": {
        "display_name": "other-agent"
      }
    }
  }
}
```

## Read registration by name

> This endpoint queries an agent registration by its display name.

Reviewer 2: ```suggestion
Fetch agent registration details by the registration display name.
```
